File: runtime-core/src/nodes/text_collector.rs
/// Text Collector Node
///
/// Accumulates streaming text tokens and yields complete sentences based on punctuation.
/// Works with streaming text generation (e.g., LFM2-Audio text output).
///
/// Pipeline flow:
///   LFM2Audio → [Text tokens] → TextCollector → [Complete sentences] → Next node
///
/// The collector receives:
/// - Individual text tokens/chunks
/// - Special tokens like <|text_end|>
///
/// When a sentence boundary is detected (. ! ? , ; :), it outputs the accumulated text.

use crate::data::RuntimeData;
use crate::error::{Error, Result};
use crate::nodes::AsyncStreamingNode;
use async_trait::async_trait;
use std::sync::Arc;
use tokio::sync::Mutex;

/// Buffer state for a single session
#[derive(Debug, Clone)]
struct TextBufferState {
    /// Accumulated text buffer
    accumulated_text: String,
    /// Total chunks accumulated
    chunks_accumulated: usize,
}

impl Default for TextBufferState {
    fn default() -> Self {
        Self {
            accumulated_text: String::new(),
            chunks_accumulated: 0,
        }
    }
}

/// Text Collector Node
pub struct TextCollectorNode {
    /// Sentence boundary characters (default: .!?,;:\n)
    boundary_chars: Vec<char>,

    /// Minimum sentence length before yielding (characters)
    min_sentence_length: usize,

    /// Yield partial sentences when <|text_end|> is received
    yield_partial_on_end: bool,

    /// Buffer states per session
    states: Arc<Mutex<std::collections::HashMap<String, TextBufferState>>>,
}

impl TextCollectorNode {
    pub fn new(
        split_pattern: Option<String>,
        min_sentence_length: Option<usize>,
        yield_partial_on_end: Option<bool>,
    ) -> Result<Self> {
        // Parse split pattern to extract boundary characters
        // Default: [.!?;\n]+ (no commas - we want to keep them in sentences)
        let boundary_chars = if let Some(pattern) = split_pattern {
            // Simple parsing: extract characters from pattern like [.!?;\n]+
            pattern.chars()
                .filter(|c| !['[', ']', '+', '\\', 'n', 'r', 't'].contains(c))
                .collect()
        } else {
            vec!['.', '!', '?', ';', '\n']
        };

        Ok(Self {
            boundary_chars,
            min_sentence_length: min_sentence_length.unwrap_or(3),
            yield_partial_on_end: yield_partial_on_end.unwrap_or(true),
            states: Arc::new(Mutex::new(std::collections::HashMap::new())),
        })
    }

    fn is_boundary_char(&self, c: char) -> bool {
        self.boundary_chars.contains(&c)
    }

    fn extract_complete_sentences(&self, buffer: &str) -> (Vec<String>, String) {
        let mut sentences = Vec::new();
        let mut current_sentence = String::new();
        let chars: Vec<char> = buffer.chars().collect();

        let mut i = 0;
        while i < chars.len() {
            let c = chars[i];
            current_sentence.push(c);

            // Check if this is a boundary character
            if self.is_boundary_char(c) {
                // Include consecutive boundary chars in the sentence
                while i + 1 < chars.len() && self.is_boundary_char(chars[i + 1]) {
                    i += 1;
                    current_sentence.push(chars[i]);
                }

                // We've found a complete sentence
                // Only trim leading whitespace, preserve trailing punctuation spacing
                let sentence = current_sentence.trim_start().to_string();
                if sentence.len() >= self.min_sentence_length {
                    sentences.push(sentence);
                }
                current_sentence.clear();
            }

            i += 1;
        }

        // Remainder (incomplete sentence) - don't trim to preserve spacing
        let remainder = current_sentence;

        (sentences, remainder)
    }
}

#[async_trait]
impl AsyncStreamingNode for TextCollectorNode {
    fn node_type(&self) -> &str {
        "TextCollectorNode"
    }

    async fn process(&self, _data: RuntimeData) -> Result<RuntimeData> {
        Err(Error::Execution(
            "TextCollectorNode requires streaming mode - use process_streaming() instead".into()
        ))
    }

    async fn process_streaming<F>(
        &self,
        data: RuntimeData,
        session_id: Option<String>,
        mut callback: F,
    ) -> Result<usize>
    where
        F: FnMut(RuntimeData) -> Result<()> + Send,
    {
        let session_key = session_id.clone().unwrap_or_else(|| "default".to_string());

        // Only process text data
        let text_chunk = match &data {
            RuntimeData::Text(text_string) => {
                text_string.clone()
            }
            _ => {
                // Pass through non-text data unchanged
                tracing::debug!("[TextCollector] Passing through non-text data: {:?}", data.data_type());
                callback(data)?;
                return Ok(1);
            }
        };

        tracing::debug!("[TextCollector] Session {}: Received text chunk: '{}'", session_key, text_chunk);

        let mut states = self.states.lock().await;
        let state = states.entry(session_key.clone()).or_insert_with(TextBufferState::default);

        // Check for special end tokens
        let has_text_end = text_chunk.contains("<|text_end|>");
        let has_audio_end = text_chunk.contains("<|audio_end|>");

        // Remove special tokens from the text (don't trim to preserve spacing)
        let cleaned_text = text_chunk
            .replace("<|text_end|>", "")
            .replace("<|audio_end|>", "")
            .to_string();

        // Add to buffer
        state.accumulated_text.push_str(&cleaned_text);
        state.chunks_accumulated += 1;

        tracing::debug!(
            "[TextCollector] Session {}: Buffer now: '{}' ({} chars)",
            session_key,
            state.accumulated_text,
            state.accumulated_text.len()
        );

        // Extract complete sentences
        let (sentences, remainder) = self.extract_complete_sentences(&state.accumulated_text);

        let mut output_count = 0;

        // Yield complete sentences
        for sentence in sentences {
            tracing::info!("[TextCollector] Session {}: Yielding sentence: '{}'", session_key, sentence);
            callback(RuntimeData::Text(sentence))?;
            output_count += 1;
        }

        // Update buffer with remainder
        state.accumulated_text = remainder;

        // Handle end tokens
        if has_text_end || has_audio_end {
            // Yield any remaining partial sentence if configured
            if self.yield_partial_on_end && !state.accumulated_text.is_empty() {
                tracing::info!(
                    "[TextCollector] Session {}: Yielding partial on end: '{}'",
                    session_key,
                    state.accumulated_text
                );
                callback(RuntimeData::Text(state.accumulated_text.clone()))?;
                output_count += 1;
                state.accumulated_text.clear();
            }

            // Pass through end tokens
            if has_text_end {
                tracing::info!("[TextCollector] Session {}: Passing through <|text_end|>", session_key);
                callback(RuntimeData::Text("<|text_end|>".to_string()))?;
                output_count += 1;
            }
            if has_audio_end {
                tracing::info!("[TextCollector] Session {}: Passing through <|audio_end|>", session_key);
                callback(RuntimeData::Text("<|audio_end|>".to_string()))?;
                output_count += 1;
            }

            // Reset state for next message
            state.accumulated_text.clear();
            state.chunks_accumulated = 0;
        }

        Ok(output_count)
    }
}

File: clients/python/remotemedia/nodes/text_collector.py
"""
Text Collector Node

Buffers incoming text tokens and yields complete sentences/patterns.
This node is useful for streaming text-to-speech where you want to accumulate
text tokens into complete sentences before passing to downstream nodes.

Key features:
- Buffers text tokens until a sentence boundary is detected
- Configurable split pattern (punctuation, newlines, etc.)
- Handles special tokens like <|text_end|> and <|audio_end|>
- Yields RuntimeData.Text for complete sentences
"""

import logging
import re
from typing import AsyncGenerator, Optional, TYPE_CHECKING

# Import RuntimeData bindings
if TYPE_CHECKING:
    from remotemedia.core.multiprocessing.data import RuntimeData

try:
    from remotemedia.core.multiprocessing.data import RuntimeData
    RUNTIME_DATA_AVAILABLE = True
except ImportError:
    RUNTIME_DATA_AVAILABLE = False
    RuntimeData = None  # type: ignore
    logging.warning("[TextCollectorNode] RuntimeData bindings not available. Using fallback implementation.")

logger = logging.getLogger(__name__)

# Configure logger
if not logger.handlers:
    console_handler = logging.StreamHandler()
    console_handler.setLevel(logging.INFO)
    formatter = logging.Formatter('%(levelname)s:%(name)s:%(message)s')
    console_handler.setFormatter(formatter)
    logger.addHandler(console_handler)
    logger.setLevel(logging.INFO)


class TextCollectorNode:
    """
    Text collector node that buffers text tokens and yields complete sentences.

    This node accumulates streaming text tokens and yields them as complete
    sentences/phrases when punctuation boundaries are detected.
    """

    def __init__(
        self,
        node_id: str,
        split_pattern: str = r'[.!?,;:\n]+',
        min_sentence_length: int = 3,
        yield_partial_on_end: bool = True,
        **kwargs
    ):
        """
        Initialize TextCollectorNode.

        Args:
            node_id: Unique identifier for this node instance
            split_pattern: Regex pattern to detect sentence boundaries (default: punctuation + newlines)
            min_sentence_length: Minimum sentence length before yielding (default: 3 chars)
            yield_partial_on_end: If True, yield partial sentences when <|text_end|> is received
        """
        self.node_id = node_id
        self.split_pattern = split_pattern
        self.min_sentence_length = min_sentence_length
        self.yield_partial_on_end = yield_partial_on_end
        self.is_streaming = True

        # Compile regex pattern for performance
        self._boundary_pattern = re.compile(split_pattern)

        logger.info(
            f"TextCollectorNode initialized: pattern='{split_pattern}', "
            f"min_length={min_sentence_length}"
        )

    async def initialize(self) -> None:
        """Initialize the node (no-op for this node)."""
        logger.info("TextCollectorNode initialized")

    async def cleanup(self) -> None:
        """Clean up the node (no-op for this node)."""
        logger.info("TextCollectorNode cleaned up")

    async def process(
        self,
        data: RuntimeData
    ) -> AsyncGenerator[RuntimeData, None]:
        """
        Process incoming text data and yield complete sentences.

        Args:
            data: RuntimeData containing text chunks (RuntimeData.Text)

        Yields:
            RuntimeData.Text for complete sentences
        """
        # Buffer for accumulating text
        text_buffer = ""

        try:
            # Validate input type
            if not data.is_text():
                logger.warning(f"TextCollectorNode expects text input, got {data.data_type()}")
                # Pass through non-text data
                yield data
                return

            # Extract text content
            text_chunk = data.as_text()
            logger.info(f"Received text chunk: '{text_chunk}'")

            # Check for special end tokens
            is_text_end = '<|text_end|>' in text_chunk
            is_audio_end = '<|audio_end|>' in text_chunk

            # Remove special tokens from the text
            cleaned_text = text_chunk
            cleaned_text = cleaned_text.replace('<|text_end|>', '')
            cleaned_text = cleaned_text.replace('<|audio_end|>', '')
            cleaned_text = cleaned_text.replace('<|[^|]+|>', '')  # Remove other special tokens

            # Add to buffer
            text_buffer += cleaned_text
            logger.info(f"Buffer now: '{text_buffer}'")

            # Find sentence boundaries in the buffer
            # We'll yield all complete sentences and keep the remainder
            sentences = []
            last_end = 0

            for match in self._boundary_pattern.finditer(text_buffer):
                # Get text up to and including the delimiter
                sentence_end = match.end()
                sentence = text_buffer[last_end:sentence_end].strip()

                # Only yield if meets minimum length
                if len(sentence) >= self.min_sentence_length:
                    sentences.append(sentence)
                    last_end = sentence_end

            # Yield complete sentences
            for sentence in sentences:
                logger.info(f"Yielding complete sentence: '{sentence}'")
                yield RuntimeData.text(sentence)

            # Update buffer to keep the remainder
            text_buffer = text_buffer[last_end:].strip()
            logger.info(f"Remaining buffer: '{text_buffer}'")

            # Handle end tokens
            if is_text_end or is_audio_end:
                # Yield any remaining partial sentence if configured
                if self.yield_partial_on_end and text_buffer:
                    logger.info(f"Yielding partial sentence on end: '{text_buffer}'")
                    yield RuntimeData.text(text_buffer)
                    text_buffer = ""

                # Pass through the end token
                if is_text_end:
                    logger.info("Yielding <|text_end|>")
                    yield RuntimeData.text("<|text_end|>")
                if is_audio_end:
                    logger.info("Yielding <|audio_end|>")
                    yield RuntimeData.text("<|audio_end|>")

        except Exception as e:
            logger.error(f"Error in TextCollectorNode: {e}", exc_info=True)
            # Yield error message
            yield RuntimeData.text(f"ERROR: Text collection failed: {str(e)}")

    def get_config(self) -> dict:
        """Get node configuration."""
        return {
            "node_id": self.node_id,
            "node_type": "TextCollectorNode",
            "split_pattern": self.split_pattern,
            "min_sentence_length": self.min_sentence_length,
            "yield_partial_on_end": self.yield_partial_on_end,
        }


# Example usage
async def main():
    """
    Example demonstrating TextCollectorNode
    """
    if not RUNTIME_DATA_AVAILABLE:
        print("RuntimeData bindings not available. Please build the Rust extension.")
        return

    import asyncio

    print("=" * 60)
    print("Text Collector Node Example")
    print("=" * 60)

    # Create collector node
    collector = TextCollectorNode(
        node_id="text_collector_1",
        split_pattern=r'[.!?,;:\n]+',
        min_sentence_length=3,
    )

    await collector.initialize()

    # Simulate streaming text tokens
    text_chunks = [
        "Hello",
        " world",
        "!",
        " How",
        " are",
        " you",
        " doing",
        " today",
        "?",
        " I'm",
        " doing",
        " great",
        ".",
        "<|text_end|>",
    ]

    print("\nSimulating streaming text tokens:")
    for chunk in text_chunks:
        print(f"  Input chunk: '{chunk}'")
        text_data = RuntimeData.text(chunk)

        async for output in collector.process(text_data):
            if output.is_text():
                text_out = output.as_text()
                print(f"  → Output: '{text_out}'")

    await collector.cleanup()
    print("\n" + "=" * 60)


if __name__ == "__main__":
    import asyncio
    asyncio.run(main())
